docs/concepts/index.md

```diff
@@ -0,0 +1,160 @@
+
+![An example HW question](./hw-question.png)
+
+#### Example `question.html`
+
+```html title="question.html"
+<pl-question-panel>
+  If $x = {{params.x}}$ and $y$ is {{params.operation}} $x$, what is $y$?
+</pl-question-panel>
+
+<pl-number-input answers-name="y" label="$y =$"></pl-number-input>
+```
+
+#### Example `server.py`
+
+```python title="server.py"
+def generate(data):
+    data["params"]["x"] = random.randint(5, 10)
+    data["params"]["operation"] = random.choice(["double", "triple"])
+
+    if data["params"]["operation"] == "double":
+        data["correct_answers"]["y"] = 2 * data["params"]["x"]
+    else:
+        data["correct_answers"]["y"] = 3 * data["params"]["x"]
+```
+
+??? example "Question with custom partial credit & feedback"
+
+    This is the same question that now gives intermediate student feedback and partial credit. This question is more complex and flexible, but this complexity is completely opt-in if you don't need it.
+
+    === "Example `question.html`"
+
+        ```html title="question.html"
+        <pl-question-panel>
+            If $x = {{params.x}}$ and $y$ is {{params.operation}} $x$, what is $y$?
+        </pl-question-panel>
+
+        <pl-number-input answers-name="y" label="$y =$"></pl-number-input>
+        <pl-submission-panel> {{feedback.y}} </pl-submission-panel>
+        ```
+
+    === "Example `server.py`"
+
+        ```python title="server.py"
+        def generate(data):
+            data["params"]["x"] = random.randint(5, 10)
+            data["params"]["operation"] = random.choice(["double", "triple"])
+
+            if data["params"]["operation"] == "double":
+                data["correct_answers"]["y"] = 2 * data["params"]["x"]
+            else:
+                data["correct_answers"]["y"] = 3 * data["params"]["x"]
+
+        def parse(data):
+            if "y" not in data["format_errors"] and data["submitted_answers"]["y"] < 0:
+                data["format_errors"]["y"] = "Negative numbers are not allowed"
+
+        def grade(data):
+            if math.isclose(data["score"], 0.0) and data["submitted_answers"]["y"] > data["params"]["x"]:
+                data["partial_scores"]["y"]["score"] = 0.5
+                data["score"] = 0.5
+                data["feedback"]["y"] = "Your value for $y$ is larger than $x$, but incorrect."
+        ```
+
+#### External Graders
+
+For coding questions, you can use an [**external grader**](../externalGrading.md) to grade a submission instead of a `server.py` file. This lets you grade student-submitted code in a sandboxed environment. There are currently external graders for C, C++, Java, Python, and more.
```

Comment:
```suggestion
For coding questions, you can use an [**external grader**](../externalGrading.md) to grade a submission instead of a `server.py` file. This lets you grade student-submitted code in a sandboxed environment. There are currently external graders for C, C++, Java, Python, and more. You can also make a custom external grader for any language.
```

Reply:
This is really nit, but: technically you could use external grader for non-coding questions. For example, I used an external grader in a question where students were expected to create a Git repository and save files in a particular format. The external grader ran a Python script with gitpython to navigate the created repository and check that all expected actions were performed (e.g., a branch was created, the new file in that branch was created, the file was added to the commit, no other files were changed, etc.). I believe we've also seen a recent case of an instructor on Slack using external graders to check datasets?

In essence, an external grader can be used in any situation where the grading process benefits from a controlled environment to be executed, and/or where grading may take more than a second to run. Coding questions are the most common case, but this could be used for other questions with complex grading (e.g., image processing, environments that require an API call, automated verification of datasets, etc.). I'm wondering if a quick mention of this flexibility may be a decent idea here? That said, this is nit, and I don't mind if you'd prefer not to waste time with these cases.